Don't leak Github Target's builder status subscription

apilloud opened the pull request:
The github status target leaks builder subscriptions on reload. This can result in a huge multiplication on API calls to github, which is ratelimited at 5000 requests per hour. This change saves off a handle to every builder status, and unsubscribes on service stop.

tardyp commented:
Did you test it on master branch?

I dont think this works like that in nine.

apilloud commented:
I didn't test this on master, only eight, but the same bug exists in both places. It looks like the mock interface the github target test implements is incomplete. I'll improve the existing test and add a new one that detects this leak.

tardyp commented:
@apilloud we would love to see this merged! will you have the time to finish it?

apilloud commented:
@tardyp Sorry, our buildbot deployment was on fire when I was working on this, so I didn't have time to write the tests. I then proceeded to forget about it. Rebased and added the test.

tardyp commented on master/buildbot/test/unit/test_status_github.py:
```diff
@@ -170,11 +170,28 @@ def test_startService(self):
+        """
+        builderstatus = FakeBuilderStatus()
+
+        result = self.status.builderAdded('builder-name', builderstatus)
```
wait what..

There is no builderStatus in buildbot nine. I guess this was a bug on eight?
github status is not yet ported to nine, so it needs to be ported first.

tardyp replied:
we really need to port this. if you are volunteer that would be very cool
see the status, so that you can find example of ported code:
http://docs.buildbot.net/latest/manual/cfg-reporters.html#githubstatus-not-migrated

apilloud replied:
The code in 'master/buildbot/status/builder.py' is just dead then? Our production and test environments are stuck on buildbot eight. Unfortunately there aren't any features that are compelling enough to push us to nine, and we've added a bunch of features on top of eight that don't port to nine, so moving isn't going to happen any time soon.
